Improve endpoint regeneration logging with detailed breakdown

tgraf opened the pull request:
* Avoid complicated defer logic to log when the policy recalculation has completed

* Simplify the logging invocations, remove duplicated log messages and use
  consistent language. Also include the build reason.

* Fix the metric accounting for policy recalculation to not include skipped
  calculations to have real insight into how long calculations take. This can
  later be improved by using labels instead.

* Provides a detailed breakdown of individual operation durations:

  ```
  msg="Completed endpoint regeneration" bpfCompilation=1.230640282s buildDuration=1.236731263s containerID=6086b396cd datapathPolicyRevision=0 desiredPolicyRevision=2 endpointID=54912 ipv4=10.11.221.144 ipv6="f00d::a0f:0:0:d680" k8sPodName=/ mapSync="215.248µs" policyCalculation=2.651822ms policyRevision=2 prepareBuild="517.631µs" proxyConfiguration="1.225µs" proxyPolicyCalculation="215.838µs" proxyWaitForAck="322.373µs" reason="syncing state to host" waitingForCTClean=422ns waitingForLock="2.069µs"
  ```

* Add DesiredPolicyRevision to endpoint logger

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5358)
<!-- Reviewable:end -->

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -648,12 +656,15 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir, reason string) (revnum uint
 
+	stats.prepareBuild.End()
+
 	if bpfHeaderfilesChanged {
```
The removal of `BuildMutex` means we no longer hold any lock while we are compiling the program for an endpoint. Is this safe? Are we relying upon the logic to ensure that only one regeneration per endpoint is taking place at a given time? : 

https://github.com/cilium/cilium/blob/d8314c7734c6efee94411bf6e506b91ac853d0e7/pkg/endpoint/policy.go#L772-L828

tgraf replied:
I checked this again. While we ensure that only one endpoint can be queued for regeneration. When it dequeued for building, parallel builds could indeed happen. I'll move the removal of the BuildMutex to a separate PR to make review easier.

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -764,6 +747,12 @@ func (e *Endpoint) regenerate(owner Owner, reason string) (retErr error) {
 	e.bumpPolicyRevisionLocked(revision)
+
+	// Set to Ready if the endpoint has not already transitioned further
+	if e.GetStateLocked() == StateRegenerating {
```
Why isn't this logic inside the `BuilderSetStateLocked`?

tgraf replied:
This clearly documents that the state may already have transitioned and that it's fine if that happened. This was absolutely not clear to me before if this is hidden in a function that is supposed to *set* the state.

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -607,14 +607,20 @@ func (e *Endpoint) regeneratePolicy(owner Owner, opts models.ConfigurationMap) (
+		"policyChanged":                  policyChanged,
+		"optionsChanged":                 optsChanged,
+		"bpfCompilationRequired":         needToRegenerateBPF,
+	}).Info("Completed endpoint policy recalculation")
```
Should this be Info or Debug?

tgraf replied:
I purposefully made it an info so we gain visibility into what is happening without having to enable debug. We had very limited visibility into how often policy was recalculated, how long it took and what the implications of the recalculation are.

aanm commented on pkg/spanstat/spanstat.go:
```diff
@@ -0,0 +1,46 @@
+
+// End ends the current span and adds the measured duration to the total
+func (s *SpanStat) End() {
+	s.totalDuration += time.Since(s.spanStart)
```
Shouldn't this be `=` instead of `+=`?

tgraf replied:
No, multiple spans can be added to the same total, e.g. waiting for locks.

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -442,12 +442,12 @@ func (e *Endpoint) removeOldRedirects(owner Owner, desiredRedirects map[string]b
+// recompiles the BPF program if needed.  Returns the implemented policy
+// revision and a boolean indicating whether a BPF compilation was required.
+//
+// Must be called with endpoint.Mutex NOT held
```
Change back documentation to say that `BuildMutex` is held.

tgraf replied:
It does not require the BuildMutex to be held. I'm not sure why the doc said that before.

raybejjani commented on pkg/endpoint/endpoint.go:
```diff
@@ -1804,11 +1805,68 @@ func (e *Endpoint) replaceIdentityLabels(l pkgLabels.Labels) int {
+func (e *Endpoint) Delete(owner Owner, releaseIP bool) []error {
 	errors := []error{}
 
+	e.UnconditionalLock()
```
Specifically in the case of delete, shouldn't BuildMutex also be held? This way we would wait for any in-process build to complete before we delete files, but we also stop any builds from starting (since once we unlock BuildMutex they would then check the state via a LockAlive and stop building?

I'm probably quite incorrect, however, since we unlock BuildMutex just before calling Delete in the one place we call it (in the old code, however, we held it).

tgraf replied:
I've moved this code out into a separate PR as it needs further work. Holding the BuildLock is not required as no build can start without before holding the endpoint lock and without verifying that the endpoint state is not in disconnecting. So it's fine to only hold the build lock while we transition the state to disconnecting.

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -567,19 +579,25 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir string, regenContext *Regene
 			e.Unlock()
 			return 0, compilationExecuted, err
 		}
+		stats.proxyPolicyCalculation.End()
```
Should this be put on line 593 if `updateNetworkPolicy` fails?

tgraf replied:
I thought about but it also gives a wrong impression on the duration as it was aborted. I also thought about closing all span brackets in `Total()`.

ianvernon replied:
Makes sense; I wasn't too partial one way or the other.

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -594,6 +612,9 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir string, regenContext *Regene
 	e.removeOldRedirects(owner, desiredRedirects, proxyWaitGroup)
+	stats.proxyConfiguration.End()
+
+	stats.prepareBuild.Start()
```
Should `End()` be invoked in the error cases as well?

ianvernon commented on pkg/endpoint/policy.go:
```diff
@@ -530,7 +522,8 @@ func (e *Endpoint) regeneratePolicy(owner Owner) (isPolicyComp bool, err error)
 			"policyRevision.repo": revision,
 			"policyChanged":       e.nextPolicyRevision > e.policyRevision,
-		}).Debug("skipping policy recalculation")
+		}).Debug("Skipping unnecessary endpoing policy recalculation")
```
endpoing --> endpoint

ianvernon reviewed (approved):
Some comments, but nothing blocking.

tgraf committed:
endpoint: Rework regeneration logging with detailed duration times

* Avoid complicated defer logic to log when the policy recalculation has completed

* Simplify the logging invocations, remove duplicated log messages and use
  consistent language. Also include the build reason.

* Fix the metric accounting for policy recalculation to not include skipped
  calculations to have real insight into how long calculations take. This can
  later be improved by using labels instead.

* Provides a detailed breakdown of individual operation durations:

  ```
  msg="Completed endpoint regeneration" bpfCompilation=1.230640282s buildDuration=1.236731263s containerID=6086b396cd datapathPolicyRevision=0 desiredPolicyRevision=2 endpointID=54912 ipv4=10.11.221.144 ipv6="f00d::a0f:0:0:d680" k8sPodName=/ mapSync="215.248µs" policyCalculation=2.651822ms policyRevision=2 prepareBuild="517.631µs" proxyConfiguration="1.225µs" proxyPolicyCalculation="215.838µs" proxyWaitForAck="322.373µs" reason="syncing state to host" waitingForCTClean=422ns waitingForLock="2.069µs"
  ```

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
endpoint: Add DesiredPolicyRevision to endpoint logger

Allows to see at what point a delta exists between desired and realized policy
revision

Signed-off-by: Thomas Graf <thomas@cilium.io>
